design/multi-factor-admin-and-step-up.md

```diff
@@ -171,145 +160,117 @@ For multi-token support in keycloak, when an authentication execution uses crede
+On the authentication execution side, classes that implement the methods that require validation of credentials will implement a new `CredentialValidator` interface, which is the following:
 ```
 package org.keycloak.authentication;
 
 import org.keycloak.credential.CredentialModel;
 import org.keycloak.credential.CredentialProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
 
 import java.util.List;
 
-public interface CredentialValidator <T extends CredentialProvider> {
-    public T getCredentialProvider(AuthenticationFlowContext context);
-    default public List<CredentialModel> getCredentials(AuthenticationFlowContext context) {
-        return context.getSession().userCredentialManager().getStoredCredentialsByType(context.getRealm(), context.getUser(), getCredentialProvider(context).getType());
+public interface CredentialValidator<T extends CredentialProvider> {
+    T getCredentialProvider(KeycloakSession session);
+    default List<CredentialModel> getCredentials(KeycloakSession session, RealmModel realm, UserModel user) {
+        return session.userCredentialManager().getStoredCredentialsByType(realm, user, getCredentialProvider(session).getType());
+    }
+    default String getType(KeycloakSession session) {
+        return getCredentialProvider(session).getType();
     }
 }
 ```
 
-And the implementation in the class would be, for `OTPFormAuthenticator` for example:
+And the implementation in the class is, for `OTPFormAuthenticator` for example:
 ```
 @Override
-public OTPCredentialProvider getCredentialProvider(AuthenticationFlowContext context) {
-    return context.getSession().getProvider(OTPCredentialProvider.class);
+public OTPCredentialProvider getCredentialProvider(KeycloakSession session) {
+    return (OTPCredentialProvider)session.getProvider(CredentialProvider.class, "keycloak-otp");
 }
 ```
-Having the associated credential provider directly in the class allows the retrieval of the credentials through the `UserCredentialManager` and the type given by the `CredentialProvider`. It also simplifies the validation code a little, as it's no longer necessary to pass by the `UserCredentialManager` to call `isValid()`, but it can be called directly from the associated `CredentialProvider`.
-
-The `Authenticator#configuredFor()` method may become obsolete with this scheme, as it is superseded by the fact that a class extends `CredentialValidator` and the list of credentials that is obtained for the user.
+Having the associated credential provider directly in the class allows the retrieval of the credentials through the `UserCredentialManager` and the type given by the `CredentialProvider`. It also simplifies the validation code a little, as it's no longer necessary to pass by the `UserCredentialManager` to call `isValid()`, but it can (and should) be called directly from the associated `CredentialProvider`.
```

Comment:
For this particular part, I personally want to doublecheck the prototype a bit more and verify if it doesn't bring any regression with relation to userStorage providers. AFAIK in current Keycloak master, it is possible to specify that your userStorage provider is able to validate certain credential types (password, otp). I am not 100% sure if this change mentioned in this paragraph and introduced in OTPAuthenticator won't introduce any limits/regressions for the usage with userStorage provider, but I hope it is ok...

I am fine with that design for now and will get back to you if I have some doubts about userStorage providers. Just wanted to let you know.